File: cashier/src/postgres_cashier.cpp
#include "postgres_cashier.hpp"
#include <pqxx/pqxx>
#include <stdexcept>

PostgresCashier::PostgresCashier(const std::string&               dsn,
                                 std::shared_ptr<IEtcdLockFactory> lock_factory,
                                 int                               pool_size)
    : pool_(dsn, pool_size)
    , lock_factory_(std::move(lock_factory))
{}

CashierResult PostgresCashier::fetch_result(pqxx::work& txn, const std::string& trader) {
    auto rows = txn.exec(
        "SELECT balance, locked FROM accounts WHERE trader_id = $1",
        pqxx::params{trader});
    if (rows.empty()) return {};
    return CashierResult{
        .balance   = rows[0][0].as<int64_t>(),
        .locked    = rows[0][1].as<int64_t>(),
        .available = rows[0][0].as<int64_t>() - rows[0][1].as<int64_t>(),
    };
}

CashierResult PostgresCashier::deposit(const std::string& trader, int64_t amount) {
    if (amount <= 0) throw std::invalid_argument("deposit amount must be > 0");

    auto lock = lock_factory_->acquire(lock_key(trader));
    auto conn = pool_.acquire();
    pqxx::work txn{*conn};
    try {
        txn.exec(
            R"(
            INSERT INTO accounts (trader_id, balance, locked, version)
            VALUES ($1, $2, 0, 0)
            ON CONFLICT (trader_id) DO UPDATE
                SET balance    = accounts.balance + $2,
                    version    = accounts.version + 1,
                    updated_at = clock_timestamp()
            )",
            pqxx::params{trader, amount});
        txn.commit();
    } catch (...) {
        txn.abort();
        throw;
    }
    // Re-read under the same lock (after commit, using a fresh txn)
    auto conn2 = pool_.acquire();
    pqxx::nontransaction rd{*conn2};
    auto row = rd.exec(
        "SELECT balance, locked FROM accounts WHERE trader_id = $1",
        pqxx::params{trader}).one_row();
    return CashierResult{
        .balance   = row[0].as<int64_t>(),
        .locked    = row[1].as<int64_t>(),
        .available = row[0].as<int64_t>() - row[1].as<int64_t>(),
    };
}

CashierResult PostgresCashier::withdraw(const std::string& trader, int64_t amount) {
    if (amount <= 0) throw std::invalid_argument("withdraw amount must be > 0");

    auto lock = lock_factory_->acquire(lock_key(trader));
    auto conn = pool_.acquire();
    pqxx::work txn{*conn};
    try {
        auto rows = txn.exec(
            "SELECT balance, locked, version FROM accounts WHERE trader_id = $1 FOR UPDATE",
            pqxx::params{trader});
        if (rows.empty()) throw UnknownTraderError("trader not found: " + trader);

        int64_t balance = rows[0][0].as<int64_t>();
        int64_t locked  = rows[0][1].as<int64_t>();
        int64_t version = rows[0][2].as<int64_t>();
        int64_t avail   = balance - locked;

        if (avail < amount)
            throw InsufficientFundsError("insufficient funds: available=" +
                                         std::to_string(avail) +
                                         " requested=" + std::to_string(amount));

        int64_t new_balance = balance - amount;
        auto updated = txn.exec(
            "UPDATE accounts SET balance = $1, version = version + 1, updated_at = clock_timestamp() "
            "WHERE trader_id = $2 AND version = $3",
            pqxx::params{new_balance, trader, version});
        if (updated.affected_rows() == 0)
            throw std::runtime_error("concurrent modification detected for trader: " + trader);

        txn.commit();
        return CashierResult{
            .balance   = new_balance,
            .locked    = locked,
            .available = new_balance - locked,
        };
    } catch (...) {
        txn.abort();
        throw;
    }
}

CashierResult PostgresCashier::escrow(const std::string& trader,
                                      uint32_t           order_id,
                                      int64_t            amount) {
    if (amount <= 0) throw std::invalid_argument("escrow amount must be > 0");

    auto lock = lock_factory_->acquire(lock_key(trader));
    auto conn = pool_.acquire();
    pqxx::work txn{*conn};
    try {
        auto rows = txn.exec(
            "SELECT balance, locked, version FROM accounts WHERE trader_id = $1 FOR UPDATE",
            pqxx::params{trader});
        if (rows.empty()) throw UnknownTraderError("trader not found: " + trader);

        int64_t balance = rows[0][0].as<int64_t>();
        int64_t locked  = rows[0][1].as<int64_t>();
        int64_t version = rows[0][2].as<int64_t>();
        int64_t avail   = balance - locked;

        if (avail < amount)
            throw InsufficientFundsError("insufficient funds: available=" +
                                         std::to_string(avail) +
                                         " requested=" + std::to_string(amount));

        int64_t new_locked = locked + amount;
        auto updated = txn.exec(
            "UPDATE accounts SET locked = $1, version = version + 1, updated_at = clock_timestamp() "
            "WHERE trader_id = $2 AND version = $3",
            pqxx::params{new_locked, trader, version});
        if (updated.affected_rows() == 0)
            throw std::runtime_error("concurrent modification detected for trader: " + trader);

        txn.exec(
            "INSERT INTO escrow_entries (order_id, trader_id, amount) VALUES ($1, $2, $3)",
            pqxx::params{order_id, trader, amount});

        txn.commit();
        return CashierResult{
            .balance   = balance,
            .locked    = new_locked,
            .available = balance - new_locked,
        };
    } catch (...) {
        txn.abort();
        throw;
    }
}

CashierResult PostgresCashier::release_escrow(uint32_t order_id) {
    // Unlocked read to discover the trader before acquiring the per-trader lock.
    // Safe: escrow rows are only deleted here or in settle(), both under the lock.
    std::string trader;
    int64_t     escrow_amount{0};
    {
        auto conn = pool_.acquire();
        pqxx::nontransaction rd{*conn};
        auto rows = rd.exec(
            "SELECT trader_id, amount FROM escrow_entries WHERE order_id = $1",
            pqxx::params{order_id});
        if (rows.empty()) throw UnknownEscrowError("escrow not found for order_id: " +
                                                    std::to_string(order_id));
        trader        = rows[0][0].as<std::string>();
        escrow_amount = rows[0][1].as<int64_t>();
    }

    auto lock = lock_factory_->acquire(lock_key(trader));
    auto conn = pool_.acquire();
    pqxx::work txn{*conn};
    try {
        auto deleted = txn.exec(
            "DELETE FROM escrow_entries WHERE order_id = $1 RETURNING trader_id, amount",
            pqxx::params{order_id});
        if (deleted.empty())
            throw UnknownEscrowError("escrow not found for order_id: " +
                                      std::to_string(order_id));

        auto rows = txn.exec(
            "SELECT balance, locked, version FROM accounts WHERE trader_id = $1 FOR UPDATE",
            pqxx::params{trader});
        if (rows.empty()) throw UnknownTraderError("trader not found: " + trader);

        int64_t balance    = rows[0][0].as<int64_t>();
        int64_t locked     = rows[0][1].as<int64_t>();
        int64_t version    = rows[0][2].as<int64_t>();
        int64_t new_locked = locked - escrow_amount;

        auto updated = txn.exec(
            "UPDATE accounts SET locked = $1, version = version + 1, updated_at = clock_timestamp() "
            "WHERE trader_id = $2 AND version = $3",
            pqxx::params{new_locked, trader, version});
        if (updated.affected_rows() == 0)
            throw std::runtime_error("concurrent modification detected for trader: " + trader);

        txn.commit();
        return CashierResult{
            .balance   = balance,
            .locked    = new_locked,
            .available = balance - new_locked,
        };
    } catch (...) {
        txn.abort();
        throw;
    }
}

CashierResult PostgresCashier::check_available(const std::string& trader) {
    // No etcd lock: advisory read. The enforcing check is inside escrow().
    auto conn = pool_.acquire();
    pqxx::nontransaction rd{*conn};
    auto rows = rd.exec(
        "SELECT balance, locked FROM accounts WHERE trader_id = $1",
        pqxx::params{trader});
    if (rows.empty()) return {};
    int64_t balance = rows[0][0].as<int64_t>();
    int64_t locked  = rows[0][1].as<int64_t>();
    return CashierResult{
        .balance   = balance,
        .locked    = locked,
        .available = balance - locked,
    };
}

File: cashier/internal/postgres_cashier.go
package internal

import (
	"context"
	"errors"
	"fmt"
	"time"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgconn"
	"github.com/jackc/pgx/v5/pgxpool"
	"google.golang.org/protobuf/proto"

	pb "github.com/secwager/secwager/cashier/gen/cashier"
)

type PostgresCashier struct {
	pool          *pgxpool.Pool
	idempotencyTTL time.Duration
}

func NewPostgresCashier(pool *pgxpool.Pool, idempotencyTTL time.Duration) *PostgresCashier {
	return &PostgresCashier{pool: pool, idempotencyTTL: idempotencyTTL}
}

func (c *PostgresCashier) Deposit(ctx context.Context, userID, idempotencyKey string, amount int64) (AccountSnapshot, error) {
	if amount <= 0 {
		return AccountSnapshot{}, fmt.Errorf("amount must be > 0")
	}
	if idempotencyKey == "" {
		return AccountSnapshot{}, fmt.Errorf("idempotency_key is required")
	}
	return c.withIdempotency(ctx, idempotencyKey, func(tx pgx.Tx) (AccountSnapshot, error) {
		_, err := tx.Exec(ctx, `
			INSERT INTO accounts (user_id, gross_balance, escrowed, version)
			VALUES ($1, $2, 0, 0)
			ON CONFLICT (user_id) DO UPDATE
				SET gross_balance = accounts.gross_balance + $2,
				    version       = accounts.version + 1,
				    updated_at    = clock_timestamp()`,
			userID, amount)
		if err != nil {
			return AccountSnapshot{}, fmt.Errorf("deposit: %w", err)
		}
		return c.fetchSnapshot(ctx, tx, userID)
	})
}

func (c *PostgresCashier) Withdraw(ctx context.Context, userID, idempotencyKey string, amount int64) (AccountSnapshot, error) {
	if amount <= 0 {
		return AccountSnapshot{}, fmt.Errorf("amount must be > 0")
	}
	if idempotencyKey == "" {
		return AccountSnapshot{}, fmt.Errorf("idempotency_key is required")
	}
	return c.withIdempotency(ctx, idempotencyKey, func(tx pgx.Tx) (AccountSnapshot, error) {
		var grossBalance, escrowed, version int64
		err := tx.QueryRow(ctx,
			`SELECT gross_balance, escrowed, version FROM accounts WHERE user_id = $1 FOR UPDATE`,
			userID).Scan(&grossBalance, &escrowed, &version)
		if errors.Is(err, pgx.ErrNoRows) {
			return AccountSnapshot{}, &UnknownUserError{Msg: "user not found: " + userID}
		}
		if err != nil {
			return AccountSnapshot{}, fmt.Errorf("withdraw select: %w", err)
		}

		if grossBalance-escrowed < amount {
			return AccountSnapshot{}, &InsufficientFundsError{
				Msg: fmt.Sprintf("insufficient funds: available=%d requested=%d", grossBalance-escrowed, amount),
			}
		}

		tag, err := tx.Exec(ctx,
			`UPDATE accounts SET gross_balance = $1, version = version + 1, updated_at = clock_timestamp()
			 WHERE user_id = $2 AND version = $3`,
			grossBalance-amount, userID, version)
		if err != nil {
			return AccountSnapshot{}, fmt.Errorf("withdraw update: %w", err)
		}
		if tag.RowsAffected() == 0 {
			return AccountSnapshot{}, fmt.Errorf("concurrent modification for user: %s", userID)
		}
		return c.fetchSnapshot(ctx, tx, userID)
	})
}

func (c *PostgresCashier) Escrow(ctx context.Context, userID string, orderID uint32, amount int64) (AccountSnapshot, error) {
	if amount <= 0 {
		return AccountSnapshot{}, fmt.Errorf("amount must be > 0")
	}

	tx, err := c.pool.Begin(ctx)
	if err != nil {
		return AccountSnapshot{}, err
	}
	defer tx.Rollback(ctx)

	var grossBalance, escrowed, version int64
	err = tx.QueryRow(ctx,
		`SELECT gross_balance, escrowed, version FROM accounts WHERE user_id = $1 FOR UPDATE`,
		userID).Scan(&grossBalance, &escrowed, &version)
	if errors.Is(err, pgx.ErrNoRows) {
		return AccountSnapshot{}, &UnknownUserError{Msg: "user not found: " + userID}
	}
	if err != nil {
		return AccountSnapshot{}, fmt.Errorf("escrow select: %w", err)
	}

	if grossBalance-escrowed < amount {
		return AccountSnapshot{}, &InsufficientFundsError{
			Msg: fmt.Sprintf("insufficient funds: available=%d requested=%d", grossBalance-escrowed, amount),
		}
	}

	newEscrowed := escrowed + amount
	tag, err := tx.Exec(ctx,
		`UPDATE accounts SET escrowed = $1, version = version + 1, updated_at = clock_timestamp()
		 WHERE user_id = $2 AND version = $3`,
		newEscrowed, userID, version)
	if err != nil {
		return AccountSnapshot{}, fmt.Errorf("escrow update: %w", err)
	}
	if tag.RowsAffected() == 0 {
		return AccountSnapshot{}, fmt.Errorf("concurrent modification for user: %s", userID)
	}

	_, err = tx.Exec(ctx,
		`INSERT INTO escrow_entries (order_id, user_id, amount) VALUES ($1, $2, $3)`,
		orderID, userID, amount)
	if err != nil {
		var pgErr *pgconn.PgError
		if errors.As(err, &pgErr) && pgErr.Code == "23505" {
			// Duplicate order_id — idempotent retry; return current snapshot.
			if rbErr := tx.Rollback(ctx); rbErr != nil {
				return AccountSnapshot{}, rbErr
			}
			return c.CheckAvailable(ctx, userID)
		}
		return AccountSnapshot{}, fmt.Errorf("escrow insert: %w", err)
	}

	snap := AccountSnapshot{GrossBalance: grossBalance, Escrowed: newEscrowed}
	if err := tx.Commit(ctx); err != nil {
		return AccountSnapshot{}, err
	}
	return snap, nil
}

func (c *PostgresCashier) ReleaseEscrow(ctx context.Context, orderID uint32) (AccountSnapshot, error) {
	// Unlocked pre-read to discover user_id and amount.
	var userID string
	var escrowAmount int64
	err := c.pool.QueryRow(ctx,
		`SELECT user_id, amount FROM escrow_entries WHERE order_id = $1`,
		orderID).Scan(&userID, &escrowAmount)
	if errors.Is(err, pgx.ErrNoRows) {
		return AccountSnapshot{}, &UnknownEscrowError{Msg: fmt.Sprintf("escrow not found for order_id: %d", orderID)}
	}
	if err != nil {
		return AccountSnapshot{}, fmt.Errorf("release_escrow pre-read: %w", err)
	}

	tx, err := c.pool.Begin(ctx)
	if err != nil {
		return AccountSnapshot{}, err
	}
	defer tx.Rollback(ctx)

	// Delete escrow entry; re-check it exists (race between two release calls).
	tag, err := tx.Exec(ctx,
		`DELETE FROM escrow_entries WHERE order_id = $1`, orderID)
	if err != nil {
		return AccountSnapshot{}, fmt.Errorf("release_escrow delete: %w", err)
	}
	if tag.RowsAffected() == 0 {
		return AccountSnapshot{}, &UnknownEscrowError{Msg: fmt.Sprintf("escrow not found for order_id: %d", orderID)}
	}

	var grossBalance, escrowed, version int64
	err = tx.QueryRow(ctx,
		`SELECT gross_balance, escrowed, version FROM accounts WHERE user_id = $1 FOR UPDATE`,
		userID).Scan(&grossBalance, &escrowed, &version)
	if err != nil {
		return AccountSnapshot{}, fmt.Errorf("release_escrow account select: %w", err)
	}

	newEscrowed := escrowed - escrowAmount
	tag, err = tx.Exec(ctx,
		`UPDATE accounts SET escrowed = $1, version = version + 1, updated_at = clock_timestamp()
		 WHERE user_id = $2 AND version = $3`,
		newEscrowed, userID, version)
	if err != nil {
		return AccountSnapshot{}, fmt.Errorf("release_escrow update: %w", err)
	}
	if tag.RowsAffected() == 0 {
		return AccountSnapshot{}, fmt.Errorf("concurrent modification for user: %s", userID)
	}

	snap := AccountSnapshot{GrossBalance: grossBalance, Escrowed: newEscrowed}
	if err := tx.Commit(ctx); err != nil {
		return AccountSnapshot{}, err
	}
	return snap, nil
}

func (c *PostgresCashier) CheckAvailable(ctx context.Context, userID string) (AccountSnapshot, error) {
	var snap AccountSnapshot
	err := c.pool.QueryRow(ctx,
		`SELECT gross_balance, escrowed FROM accounts WHERE user_id = $1`, userID).
		Scan(&snap.GrossBalance, &snap.Escrowed)
	if errors.Is(err, pgx.ErrNoRows) {
		return AccountSnapshot{}, nil
	}
	if err != nil {
		return AccountSnapshot{}, fmt.Errorf("check_available: %w", err)
	}
	return snap, nil
}

// withIdempotency wraps fn in a transaction with idempotency key deduplication.
func (c *PostgresCashier) withIdempotency(ctx context.Context, key string, fn func(pgx.Tx) (AccountSnapshot, error)) (AccountSnapshot, error) {
	tx, err := c.pool.Begin(ctx)
	if err != nil {
		return AccountSnapshot{}, err
	}
	defer tx.Rollback(ctx)

	// Lock the idempotency key row (or non-existent row) to serialize retries.
	var rawResp []byte
	var expiresAt time.Time
	err = tx.QueryRow(ctx,
		`SELECT response, expires_at FROM idempotency_keys WHERE key = $1 FOR UPDATE`, key).
		Scan(&rawResp, &expiresAt)

	if err == nil {
		// Key exists — return cached response if not expired.
		if time.Now().Before(expiresAt) {
			if rbErr := tx.Rollback(ctx); rbErr != nil {
				return AccountSnapshot{}, rbErr
			}
			var cached pb.CashierResponse
			if err := proto.Unmarshal(rawResp, &cached); err != nil {
				return AccountSnapshot{}, fmt.Errorf("unmarshal cached response: %w", err)
			}
			return AccountSnapshot{GrossBalance: cached.GrossBalance, Escrowed: cached.Escrowed}, nil
		}
		// Expired — delete and re-execute.
		if _, err := tx.Exec(ctx, `DELETE FROM idempotency_keys WHERE key = $1`, key); err != nil {
			return AccountSnapshot{}, err
		}
	} else if !errors.Is(err, pgx.ErrNoRows) {
		return AccountSnapshot{}, fmt.Errorf("idempotency select: %w", err)
	}

	snap, err := fn(tx)
	if err != nil {
		return AccountSnapshot{}, err
	}

	// Persist the response so retries get the same answer.
	encoded, err := proto.Marshal(&pb.CashierResponse{
		GrossBalance: snap.GrossBalance,
		Escrowed:     snap.Escrowed,
	})
	if err != nil {
		return AccountSnapshot{}, fmt.Errorf("marshal idempotency response: %w", err)
	}
	_, err = tx.Exec(ctx,
		`INSERT INTO idempotency_keys (key, response, expires_at) VALUES ($1, $2, $3)
		 ON CONFLICT (key) DO UPDATE SET response = $2, expires_at = $3`,
		key, encoded, time.Now().Add(c.idempotencyTTL))
	if err != nil {
		return AccountSnapshot{}, fmt.Errorf("idempotency insert: %w", err)
	}

	if err := tx.Commit(ctx); err != nil {
		return AccountSnapshot{}, err
	}
	return snap, nil
}

func (c *PostgresCashier) fetchSnapshot(ctx context.Context, tx pgx.Tx, userID string) (AccountSnapshot, error) {
	var snap AccountSnapshot
	err := tx.QueryRow(ctx,
		`SELECT gross_balance, escrowed FROM accounts WHERE user_id = $1`, userID).
		Scan(&snap.GrossBalance, &snap.Escrowed)
	if err != nil {
		return AccountSnapshot{}, fmt.Errorf("fetch snapshot: %w", err)
	}
	return snap, nil
}
